sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -1637,91 +2193,58 @@ std::unique_ptr<CurlNetworkConnection> CurlConnectionPool::ExtractOrCreateCurlCo
-
-  if (connection->IsShutdown())
-  {
-    // Can't re-used a shut down connection
-    return;
-  }
-
-  Log::Write(Logger::Level::Verbose, "Moving connection to pool...");
-
-  decltype(CurlConnectionPool::g_curlConnectionPool
-               .ConnectionPoolIndex)::mapped_type::value_type connectionToBeRemoved;
-
-  // Lock mutex to access connection pool. mutex is unlock as soon as lock is out of scope
-  std::unique_lock<std::mutex> lock(CurlConnectionPool::ConnectionPoolMutex);
-  auto& poolId = connection->GetConnectionKey();
-  auto& hostPool = g_curlConnectionPool.ConnectionPoolIndex[poolId];
-
-  if (hostPool.size() >= _detail::MaxConnectionsPerIndex && !hostPool.empty())
-  {
-    // Remove the last connection from the pool to insert this one.
-    auto lastConnection = --hostPool.end();
-    connectionToBeRemoved = std::move(*lastConnection);
-    hostPool.erase(lastConnection);
-  }
-
-  // update the time when connection was moved back to pool
-  connection->UpdateLastUsageTime();
-  hostPool.push_front(std::move(connection));
-
-  if (m_cleanThread.joinable() && !IsCleanThreadRunning)
-  {
-    // Clean thread was running before but it's finished, join it to finalize
-    m_cleanThread.join();
+#if defined(AZ_PLATFORM_LINUX)
+    if (performResult == CURLE_SSL_PEER_CERTIFICATE)
+    {
+      curl_easy_getinfo(m_handle.get(), CURLINFO_SSL_VERIFYRESULT, &result);
+      throw Http::TransportException(
+          _detail::DefaultFailedToGetNewConnectionTemplate + hostDisplayName + ". "
+          + std::string(curl_easy_strerror(performResult))
+          + ". Underlying error: " + X509_verify_cert_error_string(result));
+    }
+    else
+#endif
+    {
+      throw Http::TransportException(
+          _detail::DefaultFailedToGetNewConnectionTemplate + hostDisplayName + ". "
+          + std::string(curl_easy_strerror(performResult)));
+    }
   }
 
-  // Cleanup will start a background thread which will close abandoned connections from the pool.
-  // This will free-up resources from the app
-  // This is the only call to cleanup.
-  if (!m_cleanThread.joinable())
-  {
-    Log::Write(Logger::Level::Verbose, "Start clean thread");
-    IsCleanThreadRunning = true;
-    m_cleanThread = std::thread(CleanupThread);
-  }
-  else
+  //   Get the socket that libcurl is using from handle. Will use this to wait while
+  // reading/writing
+  // into wire
+#if defined(_MSC_VER)
+#pragma warning(push)
+// C26812: The enum type 'CURLcode' is un-scoped. Prefer 'enum class' over 'enum' (Enum.3)
+#pragma warning(disable : 26812)
+#endif
+  result = curl_easy_getinfo(m_handle.get(), CURLINFO_ACTIVESOCKET, &m_curlSocket);
+#if defined(_MSC_VER)
+#pragma warning(pop)
+#endif
+  if (result != CURLE_OK)
   {
-    Log::Write(Logger::Level::Verbose, "Clean thread running. Won't start a new one.");
+    throw Http::TransportException(
+        "Broken connection. Couldn't get the active sockect for it."
+        + std::string(curl_easy_strerror(result)));
```

Comment:
This file is now gigantic!!!  It's not that easy to know where each part of the code is.
I think it is currently grouping code for implementing:
- curl connection pool
- curl connection
- transportAdapter::Send( )
- OpenSSL ... <stuff>
- others :)

I wonder if at some point it would be better to split it into more than one only file.  (no need to be on this PR)

Reply:
To be honest, I was *seriously* considering making that change as a part of this PR.  There's a lot of mixed together goo in this file and it could absolutely be tidied up.

Reply:
Especially when I realized just how much OpenSSL goo I needed to bring in to handle CRL validation (OpenSSL seems to have somewhat inconsistent CRL validation logic and it's not easy to enable it).